### ISSUE TITLE
2.x InstanceInfo metaData modifications ignored when metaData previously set.

### ISSUE THREAD

**Issue by Crystark** [Netflix/eureka#655]
Hi,

I'm trying to keep some custom states in the  `InstanceInfo` metaData but it seems that changes are ignored in some particular cases.

```java
private final AtomicReference<InstanceInfo>	instanceInfo	= new AtomicReference<InstanceInfo>();

public EurekaRegistration(String namespace) {
	config = new ArchaiusEurekaConfig(namespace);

	String app = namespace;
	String appGroup = config.appGroup.getValue();

	instanceInfo.set(
		new InstanceInfo.Builder()
			.withId(app + "-" + PID)
			.withApp(app)
			.withAppGroup(appGroup)
			.withStatus(Status.UP)
			.withDataCenterInfo(BasicDataCenterInfo.fromSystemData())
			.build()
		);
}

public void startRegistrationClient() {
	EurekaRegistrationClient registrationClient = Eurekas.newRegistrationClientBuilder()
		.withServerResolver(ServerResolvers
			.fromHostname(config.registrationServerHost.getValue())
			.withPort(config.registrationServerPort.getValue())
		)
		.build();

	BehaviorSubject<InstanceInfo> infoSubject = BehaviorSubject.create();
	registrationClient.register(infoSubject).subscribe();

	Observable
		.interval(0, config.registrationUpdateInterval.getValue(), TimeUnit.SECONDS)
		.share()
		.subscribe(n -> {
			InstanceInfo ii = updateAndGet(iib -> iib
				.withMetaData("updated_at", String.valueOf(System.currentTimeMillis()))
			);
			infoSubject.onNext(ii);
			LOGGER.info("onNext " + ii);
		});
}

public InstanceInfo updateAndGet(Consumer<InstanceInfo.Builder> updater) {
	return instanceInfo
		.updateAndGet(ii -> {
			InstanceInfo.Builder iib = new InstanceInfo.Builder().withInstanceInfo(ii);
			updater.accept(iib);
			return iib.build();
		});
}
```

This sends a first event that appears in the `eureka2-write-server`'s log as well as in my app's logs but all next events aren't shown by the  `eureka2-write-server` although i can see those in my logs.

So i tried moving stuff about cause i had it working earlier and I got this working:
```java
private static final InstanceInfo	II	= new InstanceInfo.Builder()
	.withId("myapp-" + PID)
	.withApp("myapp")
	.withAppGroup("myapp-group")
	.withStatus(Status.UP)
	.withDataCenterInfo(BasicDataCenterInfo.fromSystemData())
	.build();

public void startRegistrationClient() {
	EurekaRegistrationClient registrationClient = Eurekas.newRegistrationClientBuilder()
		.withServerResolver(ServerResolvers
			.fromHostname(config.registrationServerHost.getValue())
			.withPort(config.registrationServerPort.getValue())
		)
		.build();

	BehaviorSubject<InstanceInfo> infoSubject = BehaviorSubject.create();
	registrationClient.register(infoSubject).subscribe();

	Observable
		.interval(0, config.registrationUpdateInterval.getValue(), TimeUnit.SECONDS)
		.share()
		.subscribe(n -> {
			InstanceInfo ii = new InstanceInfo.Builder().withInstanceInfo(II).withMetaData("updated_at", String.valueOf(System.currentTimeMillis())).build()
			infoSubject.onNext(ii);
			LOGGER.info("onNext " + ii);
		});
}
```

So I've done more testing and here's what i concluded: any new `metaData` will be ignored if the hashmap was already initialized in the previous `InstanceInfo`. Meaning: the `metaData` field needs to be `null` in the `InstanceInfo` from which you are building for it to work.

I noticed something else: If you use a referenced `metas` hashmap it won't work either.
```java
InstanceInfo ii = new InstanceInfo.Builder()
	.withInstanceInfo(instanceInfo)
	.withMetaData(metas)
	.build();
infoSubject.onNext(ii);
```

But if you create a new hashmap each time it will work
```java
InstanceInfo ii = new InstanceInfo.Builder()
	.withInstanceInfo(instanceInfo)
	.withMetaData(new HashMap<String, String>(metas))
	.build();
infoSubject.onNext(ii);
```

Then I noticed `withBuilder` actually copies the HashMap so I ended up using the following solution which works:

```java
	private final AtomicReference<InstanceInfo.Builder>	instanceInfoBuilder	= new AtomicReference<InstanceInfo.Builder>();

	public EurekaRegistration(String namespace) {
		config = new ArchaiusEurekaConfig(namespace);

		String app = namespace;
		String appGroup = config.appGroup.getValue();

		instanceInfoBuilder.set(
			new InstanceInfo.Builder()
				.withId(app + "-" + PID)
				.withApp(app)
				.withAppGroup(appGroup)
				.withStatus(Status.UP)
				.withDataCenterInfo(BasicDataCenterInfo.fromSystemData())
			);
	}

	public void startRegistrationClient() {
		EurekaRegistrationClient registrationClient = Eurekas.newRegistrationClientBuilder()
			.withServerResolver(ServerResolvers
				.fromHostname(config.registrationServerHost.getValue())
				.withPort(config.registrationServerPort.getValue())
			)
			.build();

		BehaviorSubject<InstanceInfo> infoSubject = BehaviorSubject.create();
		registrationClient.register(infoSubject).subscribe();

		Observable
			.interval(0, config.registrationUpdateInterval.getValue(), TimeUnit.SECONDS)
			.share()
			.subscribe(n -> {
				InstanceInfo.Builder ii = updateAndGet(iib -> iib.withMetaData("updated_at", String.valueOf(System.currentTimeMillis())));
				infoSubject.onNext(ii.build());
				LOGGER.info("emited: " + ii);
			});
	}

	public InstanceInfo.Builder updateAndGet(Consumer<InstanceInfo.Builder> updater) {
		return instanceInfoBuilder
			.updateAndGet(iib -> {
				InstanceInfo.Builder iib2 = new InstanceInfo.Builder()
					.withBuilder(iib);
				updater.accept(iib2);
				return iib2;
			});
	}
```

**Comment by qiangdavidliu** [comment]
@Crystark you are right, thanks for raising the issue. The problem appears to be in the withInstanceInfo() method inside the builder, where instead of copying the metadata map, it assigns to the older instance's metadata map.

**Comment by qiangdavidliu** [comment]
Fixed in #658
